### PR TITLE
feat: import CSV UI

### LINKED ISSUE #6772: Feature: Import CSV/TSV files
Anki Desktop allows the import of CSV/TSV, and makes use of the auto-detection of CSV files which is built into Python.

Splintering off this task from #3517 as this is close to complete locally.

There will likely be follow up support tickets which would best be associated with a specific ticket rather than an 'Epic'. This is a significant port of Python and C code, and there is a large likelihood for bugs.

### PR DESCRIPTION
**Pull request by david-allison**
## Purpose / Description
We want to allow the user to import CSVs

## Fixes
Fixes #6772 

## Approach

We already have the python code converted. We use a fragment-based approach for the UI

* Host Fragment
  * Load File
  * Options
    * Delimiter Dialog
    * Mapping (CSV -> Field) Fragment
  * Load/Display Log

⚠️ We use Kotlin coroutines for async, mostly as I'd like to try them out as a new async mechanism, I'm very open to changing this if it's not ideal
* Parcelize has been added to reduce the code needed to create bundles

## How Has This Been Tested?

Some unit tests, tested on my phone

⚠️ Large feature, likely needs more integration tests on the progress/end state

## Checklist
- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [x] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)
  - [ ] ❌See below

<details><summary>Screenshots</summary>

image
image
image
image
image
image

</details>

<details><summary>Accessibility failures</summary>

image
image
image
image
image
image

</details>

### CODE REVIEW

**Comment by david-allison-1** [comment]
Blocked on #9423

**Comment by mikehardy** [comment]
#9423 is in!

**Comment by david-allison-1** [comment]
I expect this'll pass CI now. Will remove the draft if it does

Review questions: 

* should we backup before import? Will this potentially make restoration less likely (if a user messes up, I feel they're likely to try again a few times)
* How do we we want to handle the accessibility issues?

File: AnkiDroid/src/main/java/com/ichi2/anki/AnkiActivity.java [hunk 1]
```diff
@@ -654,6 +660,9 @@ protected boolean showedActivityFailedScreen(Bundle savedInstanceState) {
         Timber.w("Activity started with no application instance");
         UIUtils.showThemedToast(this, getString(R.string.ankidroid_cannot_open_after_backup_try_again), false);
 
+        // fixes: java.lang.IllegalStateException: You need to use a Theme.AppCompat theme (or descendant) with this activity.
+        // on Importer
+        Themes.setTheme(this);
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/AnkiActivity.java:665]: this needs an explicit re-test

File: AnkiDroid/src/main/java/com/ichi2/anki/importer/CsvMapping.kt [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+            return
+        }
+
+        // If the mapping already exists: it can only be added once, remove it.
+        if (optionToSelect !is MapToNothing) {
+            val index = internalMapping.indexOf(optionToSelect)
+            if (index != -1) {
+                internalMapping[index] = MapToNothing
+            }
+            internalMapping.remove(optionToSelect)
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/importer/CsvMapping.kt:87]: I don't think this is necessary given the lines above

File: AnkiDroid/src/main/java/com/ichi2/anki/importer/ImporterFileSelectionFragment.kt [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+            UIUtils.showThemedToast(context, getString(R.string.import_error_exception, e.localizedMessage), false)
+            return false
+        }
+        return true
+    }
+
+    private fun copyToCache(result: Uri): Result<CacheFilePath> {
+        return kotlin.runCatching {
+            val inputStream = requireContext().contentResolver.openInputStream(result)
+            val outputPath = File.createTempFile("import", ".tmp").absolutePath
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/importer/ImporterFileSelectionFragment.kt:117]: AnkiDroid typically doesn't handle clearing the cache, and this seems normal for Android apps, do we want to continue with this?

File: AnkiDroid/src/main/java/com/ichi2/anki/importer/ImporterImportProgressFragment.kt [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+    private lateinit var options: ImportOptions
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+
+        col = CollectionHelper.getInstance().getCol(requireContext())
+
+        options = requireArguments().getParcelable("options")!!
+
+        Timber.d("options: %s", options)
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/importer/ImporterImportProgressFragment.kt:64]: this logs file paths (as a d), is this acceptable?

File: AnkiDroid/src/main/java/com/ichi2/anki/importer/ImporterImportProgressFragment.kt [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+    }
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+
+        downloadPercentage = view.findViewById(R.id.import_progress_text)
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/importer/ImporterImportProgressFragment.kt:70]: How would you feel about using a DI framework to reduce code here?

File: AnkiDroid/src/main/java/com/ichi2/anki/importer/ImporterImportProgressFragment.kt [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+        importProgressDone = view.findViewById(R.id.import_progress_done)
+
+        Timber.i("Starting import process")
+
+        val exceptionHandler = CoroutineExceptionHandler { _, ex ->
+            Timber.w(ex, "Importing failed")
+            AnkiDroidApp.sendExceptionReport(ex, "import failed")
+            onException(ex)
+        }
+        GlobalScope.launch(Dispatchers.Main + exceptionHandler) {
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/importer/ImporterImportProgressFragment.kt:84]: @Arthur-Milchior pinging for your opinion on me testing a new async mechanism, and thoughts on whether this API is too far from what we're currently using

File: AnkiDroid/src/main/java/com/ichi2/anki/importer/ImporterOptionSelectionFragment.kt [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+        /** The key identifying the fragment result */
+        const val RESULT_KEY = "ImportOptions"
+        /**
+         * The key of the options in the resulting bundle
+         * This is an extra of type [ImportOptions]
+         */
+        const val RESULT_BUNDLE_OPTIONS = "importOptions"
+
+        /**
+         *
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/importer/ImporterOptionSelectionFragment.kt:320]: missing comment

File: AnkiDroid/src/main/java/com/ichi2/anki/importer/ImporterImportProgressFragment.kt (hunk 4 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/anki/importer/ImporterImportProgressFragment.kt:64]: "as a d" what do you mean?

File: AnkiDroid/src/main/java/com/ichi2/anki/importer/ImporterImportProgressFragment.kt (hunk 5 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/anki/importer/ImporterImportProgressFragment.kt:70]: I had to Google it, and I see no issue here.
I mean, I'm never totally fan of adding state, but I'm not fan of huge method either

File: AnkiDroid/src/main/java/com/ichi2/anki/importer/ImporterImportProgressFragment.kt (hunk 6 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/anki/importer/ImporterImportProgressFragment.kt:84]: I don't know the first thing about GlobalScope.launch, so hard to tell.
My main concern with switching with a new async mechanism is that all async code currently assume that a single background thread runs at one point in time (appart from synchronization thread). In particular, it means that there is never two activities touching the database at the same time.
I suppose that "performImport" will write to the database.
I can't imagine any actual trouble, but even if it works as expected, still not a great idea to break this invariant until we know how we want async to evolve

Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/anki/importer/ImporterImportProgressFragment.kt:84]: I don't know the first thing about GlobalScope.launch, so hard to tell.
My main concern with switching with a new async mechanism is that all async code currently assume that a single background thread runs at one point in time (appart from synchronization thread). In particular, it means that there is never two activities touching the database at the same time.
I suppose that "performImport" will write to the database.
I can't imagine any actual trouble, but even if it works as expected, still not a great idea to break this invariant until we know how we want async to evolve.

Plus, I'd assume it make testing it more complex as you can't use the same tooling to move this method to frontend

File: AnkiDroid/src/main/java/com/ichi2/anki/importer/ImporterImportProgressFragment.kt (hunk 4 above)
Comment (david-allison-1) [reply AnkiDroid/src/main/java/com/ichi2/anki/importer/ImporterImportProgressFragment.kt:64]: `Timber.d`

**Comment by david-allison-1** [comment]
de-conflicted and rebased

Changes to libAnki were (correctly!) breaking in Kotlin, but not for Java (some methods were nullable, which weren't previously annotated)

**Comment by mindinsomnia** [comment]
Import-Workflow
UI Mockup for CSV Import Workflow:

Rationale:

**Menu -> Import Cards:** Clearly labelled now so users should be able to tell easily from that menu item that if they desire to import some cards, to follow that pathway.

**Import Cards screen:** Clearly explains the process they are about to begin, both informs the user what type of files are supported for importing, and allows the user to choose the format to import and begin a process. Followed by a file selection screen which should be self explanatory at that point.

**Select Deck:** Button to popup what I suspect would be some kind of chooser dialog for choosing a deck, displaying selected deck gives the user a chance to preview that choice before they continue to avoid mistakes.

**CSV Data:** Preview the data to ensure it's being formatted nicely before continuing, chance to choose one of several common CSV delimiters if the current automatically selected delimiter is resulting in incorrect formatting. Preview does not need to list all data, even the first 100 or so rows would be sufficient.

**Import Options:** All the other decisions and choices out of the way, the user can focus on how the data is mapped, and confirm the operation to begin importing.

**Summary Page:** Neatly displays results of import process.

**Comment by david-allison-1** [comment]
Soft-blocked on https://github.com/ankidroid/Anki-Android/pull/9466

File: AnkiDroid/src/main/java/com/ichi2/anki/importer/ImporterImportProgressFragment.kt (hunk 4 above)
Comment (mikehardy) [reply AnkiDroid/src/main/java/com/ichi2/anki/importer/ImporterImportProgressFragment.kt:64]: seems fine to me, especially as a debug log both from perf or security perspective

File: AnkiDroid/src/main/java/com/ichi2/anki/importer/ImporterImportProgressFragment.kt (hunk 6 above)
Comment (mikehardy) [reply AnkiDroid/src/main/java/com/ichi2/anki/importer/ImporterImportProgressFragment.kt:84]: Agreed in general that testing a new async exec mechanism as a sort of one-off mixed in another PR doesn't seem like the right spot.

**Comment by mikehardy** [comment]
mostly just noting soft-block PR was merged, and I took a moment to briefly scan comments and perform Seagull Management duties (swoop in, squawk, leave droppings, swoop out hahah)

**Comment by david-allison-1** [comment]
Not going to get to this for a while - scoped storage needs to happen

**Comment by david-allison** [comment]
Superseded by https://github.com/ankidroid/Anki-Android/pull/12105

**Pull request closed without merging**